File: rtsp/src/main/java/com/pedro/rtsp/rtcp/BaseSenderReport.java
package com.pedro.rtsp.rtcp;

import android.util.Log;
import com.pedro.rtsp.rtsp.Protocol;
import com.pedro.rtsp.rtsp.RtpFrame;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Random;

/**
 * Created by pedro on 7/11/18.
 */

public abstract class BaseSenderReport {

  protected static final String TAG = "BaseSenderReport";
  protected static final int PACKET_LENGTH = 28;
  private static final int MTU = 1500;
  private final long interval = 3000;

  private final byte[] videoBuffer = new byte[MTU];
  private final byte[] audioBuffer = new byte[MTU];

  private long videoTime;
  private long audioTime;
  private int videoPacketCount;
  private int videoOctetCount;
  private int audioPacketCount;
  private int audioOctetCount;

  BaseSenderReport() {
    /*							     Version(2)  Padding(0)					 					*/
    /*									 ^		  ^			PT = 0	    						*/
    /*									 |		  |				^								*/
    /*									 | --------			 	|								*/
    /*									 | |---------------------								*/
    /*									 | ||													*/
    /*									 | ||													*/
    videoBuffer[0] = (byte) Integer.parseInt("10000000", 2);
    audioBuffer[0] = (byte) Integer.parseInt("10000000", 2);

    /* Packet Type PT */
    videoBuffer[1] = (byte) 200;
    audioBuffer[1] = (byte) 200;

    /* Byte 2,3          ->  Length		                     */
    setLong(videoBuffer, PACKET_LENGTH / 4 - 1, 2, 4);
    setLong(audioBuffer, PACKET_LENGTH / 4 - 1, 2, 4);
    /* Byte 4,5,6,7      ->  SSRC                            */
    setLong(videoBuffer, new Random().nextInt(), 4, 8);
    setLong(audioBuffer, new Random().nextInt(), 4, 8);
    /* Byte 8,9,10,11    ->  NTP timestamp hb				 */
    /* Byte 12,13,14,15  ->  NTP timestamp lb				 */
    /* Byte 16,17,18,19  ->  RTP timestamp		             */
    /* Byte 20,21,22,23  ->  packet count				 	 */
    /* Byte 24,25,26,27  ->  octet count			         */
  }

  public static BaseSenderReport getInstance(Protocol protocol, int videoSourcePort,
      int audioSourcePort) {
    return protocol == Protocol.TCP ? new SenderReportTcp()
        : new SenderReportUdp(videoSourcePort, audioSourcePort);
  }

  public abstract void setDataStream(OutputStream outputStream, String host);

  public void update(RtpFrame rtpFrame) {
    if (rtpFrame.getChannelIdentifier() == (byte) 2) {
      updateVideo(rtpFrame);
    } else {
      updateAudio(rtpFrame);
    }
  }

  public abstract void sendReport(byte[] buffer, RtpFrame rtpFrame, String type, int packetCount,
      int octetCount) throws IOException;

  private void updateVideo(RtpFrame rtpFrame) {
    videoPacketCount++;
    videoOctetCount += rtpFrame.getLength();
    setLong(videoBuffer, videoPacketCount, 20, 24);
    setLong(videoBuffer, videoOctetCount, 24, 28);
    if (System.currentTimeMillis() - videoTime >= interval) {
      videoTime = System.currentTimeMillis();
      setData(videoBuffer, System.nanoTime(), rtpFrame.getTimeStamp());
      try {
        sendReport(videoBuffer, rtpFrame, "Video", videoPacketCount, videoOctetCount);
      } catch (IOException e) {
        Log.e(TAG, "Error", e);
      }
    }
  }

  private void updateAudio(RtpFrame rtpFrame) {
    audioPacketCount++;
    audioOctetCount += rtpFrame.getLength();
    setLong(audioBuffer, audioPacketCount, 20, 24);
    setLong(audioBuffer, audioOctetCount, 24, 28);
    if (System.currentTimeMillis() - audioTime >= interval) {
      audioTime = System.currentTimeMillis();
      setData(audioBuffer, System.nanoTime(), rtpFrame.getTimeStamp());
      try {
        sendReport(audioBuffer, rtpFrame, "Audio", audioPacketCount, audioOctetCount);
      } catch (IOException e) {
        Log.e(TAG, "Error", e);
      }
    }
  }

  public void reset() {
    videoPacketCount = videoOctetCount = 0;
    audioPacketCount = audioOctetCount = 0;
    videoTime = audioTime = 0;
    setLong(videoBuffer, videoPacketCount, 20, 24);
    setLong(videoBuffer, videoOctetCount, 24, 28);
    setLong(audioBuffer, audioPacketCount, 20, 24);
    setLong(audioBuffer, audioOctetCount, 24, 28);
  }

  public abstract void close();

  private void setLong(byte[] buffer, long n, int begin, int end) {
    for (end--; end >= begin; end--) {
      buffer[end] = (byte) (n % 256);
      n >>= 8;
    }
  }

  private void setData(byte[] buffer, long ntpts, long rtpts) {
    long hb = ntpts / 1000000000;
    long lb = ((ntpts - hb * 1000000000) * 4294967296L) / 1000000000;
    setLong(buffer, hb, 8, 12);
    setLong(buffer, lb, 12, 16);
    setLong(buffer, rtpts, 16, 20);
  }
}

File: rtsp/src/main/java/com/pedro/rtsp/rtcp/BaseSenderReport.kt
/*
 * Copyright (C) 2024 pedroSG94.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pedro.rtsp.rtcp

import com.pedro.common.TimeUtils
import com.pedro.common.socket.base.SocketType
import com.pedro.common.socket.base.StreamSocket
import com.pedro.common.socket.base.TcpStreamSocket
import com.pedro.rtsp.rtsp.Protocol
import com.pedro.rtsp.rtsp.RtpFrame
import com.pedro.rtsp.utils.RtpConstants
import com.pedro.rtsp.utils.setLong
import java.io.IOException

/**
 * Created by pedro on 7/11/18.
 */
abstract class BaseSenderReport internal constructor() {

  private val interval: Long = 3000
  private val videoBuffer = ByteArray(RtpConstants.REPORT_PACKET_LENGTH)
  private val audioBuffer = ByteArray(RtpConstants.REPORT_PACKET_LENGTH)
  private var videoTime: Long = 0
  private var audioTime: Long = 0
  private var videoPacketCount = 0L
  private var videoOctetCount = 0L
  private var audioPacketCount = 0L
  private var audioOctetCount = 0L

  companion object {
    @JvmStatic
    fun getInstance(
      socketType: SocketType,
      protocol: Protocol, host: String,
      videoSourcePort: Int?, audioSourcePort: Int?,
      videoServerPort: Int?, audioServerPort: Int?,
    ): BaseSenderReport {
      return if (protocol === Protocol.TCP) {
        SenderReportTcp()
      } else {
        val videoSocket = if (videoServerPort != null) {
          StreamSocket.createUdpSocket(socketType, host, videoServerPort, videoSourcePort)
        } else null
        val audioSocket = if (audioServerPort != null) {
          StreamSocket.createUdpSocket(socketType, host, audioServerPort, audioSourcePort)
        } else null
        SenderReportUdp(videoSocket, audioSocket)
      }
    }
  }

  init {
    /*							     Version(2)  Padding(0)					 					*/
    /*									 ^		  ^			PT = 0	    						*/
    /*									 |		  |				^								*/
    /*									 | --------			 	|								*/
    /*									 | |---------------------								*/
    /*									 | ||													*/
    /*									 | ||													*/
    videoBuffer[0] = 0x80.toByte()
    audioBuffer[0] = 0x80.toByte()

    /* Packet Type PT */
    videoBuffer[1] = 200.toByte()
    audioBuffer[1] = 200.toByte()

    /* Byte 2,3          ->  Length		                     */
    videoBuffer.setLong(RtpConstants.REPORT_PACKET_LENGTH / 4 - 1L, 2, 4)
    audioBuffer.setLong(RtpConstants.REPORT_PACKET_LENGTH / 4 - 1L, 2, 4)
    /* Byte 4,5,6,7      ->  SSRC                            */
    /* Byte 8,9,10,11    ->  NTP timestamp hb				 */
    /* Byte 12,13,14,15  ->  NTP timestamp lb				 */
    /* Byte 16,17,18,19  ->  RTP timestamp		             */
    /* Byte 20,21,22,23  ->  packet count				 	 */
    /* Byte 24,25,26,27  ->  octet count			         */
  }

  fun setSSRC(ssrcVideo: Long, ssrcAudio: Long) {
    videoBuffer.setLong(ssrcVideo, 4, 8)
    audioBuffer.setLong(ssrcAudio, 4, 8)
  }

  @Throws(IOException::class)
  abstract suspend fun setSocket(socket: TcpStreamSocket)

  @Throws(IOException::class)
  suspend fun update(rtpFrame: RtpFrame): Boolean {
    return if (rtpFrame.channelIdentifier == RtpConstants.trackVideo) {
      updateVideo(rtpFrame)
    } else {
      updateAudio(rtpFrame)
    }
  }

  @Throws(IOException::class)
  abstract suspend fun sendReport(buffer: ByteArray, rtpFrame: RtpFrame)

  @Throws(IOException::class)
  private suspend fun updateVideo(rtpFrame: RtpFrame): Boolean {
    videoPacketCount++
    videoOctetCount += rtpFrame.length
    videoBuffer.setLong(videoPacketCount, 20, 24)
    videoBuffer.setLong(videoOctetCount, 24, 28)
    if (TimeUtils.getCurrentTimeMillis() - videoTime >= interval) {
      videoTime = TimeUtils.getCurrentTimeMillis()
      setData(videoBuffer, TimeUtils.getCurrentTimeNano(), rtpFrame.timeStamp)
      sendReport(videoBuffer, rtpFrame)
      return true
    }
    return false
  }

  @Throws(IOException::class)
  private suspend fun updateAudio(rtpFrame: RtpFrame): Boolean {
    audioPacketCount++
    audioOctetCount += rtpFrame.length
    audioBuffer.setLong(audioPacketCount, 20, 24)
    audioBuffer.setLong(audioOctetCount, 24, 28)
    if (TimeUtils.getCurrentTimeMillis() - audioTime >= interval) {
      audioTime = TimeUtils.getCurrentTimeMillis()
      setData(audioBuffer, TimeUtils.getCurrentTimeNano(), rtpFrame.timeStamp)
      sendReport(audioBuffer, rtpFrame)
      return true
    }
    return false
  }

  fun reset() {
    videoOctetCount = 0
    videoPacketCount = 0
    audioOctetCount = 0
    audioPacketCount = 0
    audioTime = 0
    videoTime = 0
    videoBuffer.setLong(videoPacketCount, 20, 24)
    videoBuffer.setLong(videoOctetCount, 24, 28)
    audioBuffer.setLong(audioPacketCount, 20, 24)
    audioBuffer.setLong(audioOctetCount, 24, 28)
  }

  abstract suspend fun close()

  private fun setData(buffer: ByteArray, ntpts: Long, rtpts: Long) {
    val hb = ntpts / 1000000000
    val lb = (ntpts - hb * 1000000000) * 4294967296L / 1000000000
    buffer.setLong(hb, 8, 12)
    buffer.setLong(lb, 12, 16)
    buffer.setLong(rtpts, 16, 20)
  }
}